hazelcast/src/main/java/com/hazelcast/map/impl/iterator/MapIterator.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.iterator;
+
+import com.hazelcast.map.impl.proxy.MapProxyImpl;
+
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Iterator for iterating map entries in the whole cluster. The values
+ * are fetched in batches.
+ * It uses {@link MapPartitionIterator} and provides the same guarantees
+ * with this partition iterator.
+ *
+ * @see MapPartitionIterator
+ */
+public class MapIterator<K, V> extends AbstractMapIterator<K, V> {
+
+    public MapIterator(MapProxyImpl<K, V> mapProxy, int fetchSize, int partitionCount, boolean prefetchValues) {
+        super(IntStream.range(0, partitionCount).boxed()
+                .map(partitionId -> mapProxy.iterator(fetchSize, partitionId, prefetchValues))
```

Comment:
Looks like here might be a potential performance problem: each `mapProxy.iterator(...)` call creates an iterator and all the iterators are created at once. If I understand the code correctly (see `MapPartitionIterator` constructor and `AbstractMapPartitionIterator.advance`), each such iterator fetches and stores `fetchSize` entires during the construction. So `partition_count * fetch_size` entries are stored by all the iterators as a whole, with the current defaults that's `271 * 100 = 27 100` entries, with a fetch size of `1000` that's `271 000` entries.

Probably a better solution would be keeping just a single iterator instance active, once it's exhausted it can be replaced with an iterator for the next partition. That would also amortize the initial delay associated with the iterators construction: currently `partition_count` fetch operations are issued sequentially during the construction and that may introduce a noticeable delay before the map iterator become available for the actual iteration.